In [ ]:
import pandas as pd
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sbn
from altair import Chart, X, Y, Color, Scale
import altair as alt
from vega_datasets import data
import requests
from bs4 import BeautifulSoup

In [ ]:
sl = pd.read_csv('shot_logs.csv')

### link to data: https://www.kaggle.com/dansbecker/nba-shot-logs

In [ ]:
alt.data_transformers.disable_max_rows()

DataTransformerRegistry.enable('default')

# Questions

Who are the best shooters in the league? <br>
What can we do to counter these players?<br>
Who are the best defenders to stop these players?<br>
Should we move specific defenders to stop these shooters?<br>

In [ ]:
ms = sl[sl.SHOT_RESULT == 'made']

# The best Shooters in the league

In [ ]:
list_of_players = []
newdict ={}
for i in sl.player_name.unique():
   list_of_players.append(i)
#print(list_of_players)
for i in list_of_players:
    temp = ms[ms.player_name == i]
    #print(i)
    #print(len(temp))
    newdict[i] = len(temp)
a = sorted(newdict.items(), key=lambda x: x[1],reverse=True)
print(a[:10])
#top10players = ['nikola vucevic', 'lebron james', 'james harden', 'mnta ellis', 'lamarcus aldridge', 'stephen curry', 'anthony davis', 'klay thompson', 'blake griffin', 'kyrie irving']
#print(top10players)

[('nikola vucevic', 480), ('lebron james', 478), ('james harden', 474), ('mnta ellis', 473), ('lamarcus aldridge', 473), ('stephen curry', 470), ('anthony davis', 457), ('klay thompson', 449), ('blake griffin', 447), ('kyrie irving', 439)]


# Nikola Vucevic shot data

In [ ]:
nv = ms[ms.player_name == 'nikola vucevic']

chart1 = alt.Chart(nv).mark_bar().encode(
    x=X('SHOT_DIST', bin=True),
    y='count()',
).properties(
    height=300,
    width=300
)

chart2 = alt.Chart(nv).mark_bar().encode(
    x=X('DRIBBLES', bin=True),
    y='count()'
).properties(
    height=300,
    width=300
)


chart3 = alt.Chart(nv).mark_bar().encode(
    x=X('CLOSE_DEF_DIST', bin=True),
    y='count()'
).properties(
    height=300,
    width=300
)
chart1 | chart2 | chart3

alt.HConcatChart(...)

### Nikola Vucevic likes to take shots close to the hoop, with little to no movement, and with a defender very close to him. If you were defending against Nikola, you would want to keep him far away from the hoop and make him move by playing very aggressive.

In [ ]:
missedshot = sl[sl.SHOT_RESULT == 'missed']

In [ ]:
missedshot

GAME_ID                   MATCHUP LOCATION  W  FINAL_MARGIN  \
1       21400899  MAR 04, 2015 - CHA @ BKN        A  W            24   
2       21400899  MAR 04, 2015 - CHA @ BKN        A  W            24   
3       21400899  MAR 04, 2015 - CHA @ BKN        A  W            24   
4       21400899  MAR 04, 2015 - CHA @ BKN        A  W            24   
5       21400899  MAR 04, 2015 - CHA @ BKN        A  W            24   
...          ...                       ...      ... ..           ...   
128059  21400033  NOV 01, 2014 - BKN @ DET        A  W            12   
128060  21400006  OCT 29, 2014 - BKN @ BOS        A  L           -16   
128061  21400006  OCT 29, 2014 - BKN @ BOS        A  L           -16   
128064  21400006  OCT 29, 2014 - BKN @ BOS        A  L           -16   
128067  21400006  OCT 29, 2014 - BKN @ BOS        A  L           -16   

        SHOT_NUMBER  PERIOD GAME_CLOCK  SHOT_CLOCK  DRIBBLES  ...  SHOT_DIST  \
1                 2       1       0:14         3.4         0  ...       28.2   
2                 3       1       0:00         NaN         3  ...       10.1   
3                 4       2      11:47        10.3         2  ...       17.2   
4                 5       2      10:34        10.9         2  ...        3.7   
5                 6       2       8:15         9.1         2  ...       18.4   
...             ...     ...        ...         ...       ...  ...        ...   
128059            4       4       8:34        19.8         0  ...       22.7   
128060            1       1       1:59        11.4        16  ...       12.6   
128061            2       2      10:10        19.0         0  ...        7.4   
128064            5       3       1:52        18.3         5  ...        8.7   
128067            8       4       2:37         9.1         4  ...       18.3   

        PTS_TYPE  SHOT_RESULT   CLOSEST_DEFENDER CLOSEST_DEFENDER_PLAYER_ID  \
1              3       missed  Bogdanovic, Bojan                     202711   
2              2       missed  Bogdanovic, Bojan                     202711   
3              2       missed      Brown, Markel                     203900   
4              2       missed    Young, Thaddeus                     201152   
5              2       missed    Williams, Deron                     101114   
...          ...          ...                ...                        ...   
128059         3       missed     Augustin, D.J.                     201571   
128060         2       missed       Rondo, Rajon                     200765   
128061         2       missed     Bradley, Avery                     202340   
128064         2       missed      Smart, Marcus                     203935   
128067         2       missed     Bradley, Avery                     202340   

        CLOSE_DEF_DIST  FGM  PTS    player_name player_id  
1                  6.1    0    0  brian roberts    203148  
2                  0.9    0    0  brian roberts    203148  
3                  3.4    0    0  brian roberts    203148  
4                  1.1    0    0  brian roberts    203148  
5                  2.6    0    0  brian roberts    203148  
...                ...  ...  ...            ...       ...  
128059             4.0    0    0   jarrett jack    101127  
128060             4.8    0    0   jarrett jack    101127  
128061             2.7    0    0   jarrett jack    101127  
128064             0.8    0    0   jarrett jack    101127  
128067             3.0    0    0   jarrett jack    101127  

[70164 rows x 21 columns]

In [ ]:
list_of_defenders = []
for i in sl.CLOSEST_DEFENDER.unique():
   list_of_defenders.append(i)

for i in list_of_defenders:
    temp = missedshot[missedshot.player_name == i]
    #print(i)
    #print(len(temp))
    newdict[i] = len(temp)
a = sorted(newdict.items(), key=lambda x: x[1],reverse=True)
print(a[:10])



[('nikola vucevic', 480), ('lebron james', 478), ('james harden', 474), ('mnta ellis', 473), ('lamarcus aldridge', 473), ('stephen curry', 470), ('anthony davis', 457), ('klay thompson', 449), ('blake griffin', 447), ('kyrie irving', 439)]


## Klay Thompson defensive stats

In [ ]:
nv = missedshot[missedshot.CLOSEST_DEFENDER == 'Thompson, Klay']

chart1 = alt.Chart(nv).mark_bar().encode(
    x=X('SHOT_DIST', bin=True),
    y='count()',
).properties(
    height=300,
    width=300
)

chart2 = alt.Chart(nv).mark_bar().encode(
    x=X('DRIBBLES', bin=True),
    y='count()'
).properties(
    height=300,
    width=300
)


chart3 = alt.Chart(nv).mark_bar().encode(
    x=X('CLOSE_DEF_DIST', bin=True),
    y='count()'
).properties(
    height=300,
    width=300
)
chart1 | chart2 | chart3

alt.HConcatChart(...)

In [ ]:
nv.mean(axis=0)

<ipython-input-15-9df0fb3fe1ce>:1: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  nv.mean(axis=0)


GAME_ID                       2.140043e+07
FINAL_MARGIN                 -1.021967e+01
SHOT_NUMBER                   8.780328e+00
PERIOD                        2.609836e+00
SHOT_CLOCK                    1.199178e+01
DRIBBLES                      2.311475e+00
TOUCH_TIME                    2.979672e+00
SHOT_DIST                     1.738459e+01
PTS_TYPE                      2.413115e+00
CLOSEST_DEFENDER_PLAYER_ID    2.026910e+05
CLOSE_DEF_DIST                4.121967e+00
FGM                           0.000000e+00
PTS                           0.000000e+00
player_id                     1.594867e+05
dtype: float64

dribbles avg: 2.3 <br>
shot distance: 17.3 <br>
avg distance to shooter: 4.12 <br>
as a defender Klay likes to keep shooters far away from the hoop and close to him, he also likes to stop them from moving

In [ ]:
nvms = sl[sl.player_name == 'nikola vucevic']

In [ ]:
nvms = nvms[nvms.CLOSEST_DEFENDER == 'Thompson, Klay']

In [ ]:
nvms

GAME_ID                     MATCHUP LOCATION  W  FINAL_MARGIN  \
43941  21400212  NOV 26, 2014 - ORL vs. GSW        H  L           -15   

       SHOT_NUMBER  PERIOD GAME_CLOCK  SHOT_CLOCK  DRIBBLES  ...  SHOT_DIST  \
43941           12       3       3:02        24.0         0  ...        3.4   

       PTS_TYPE  SHOT_RESULT CLOSEST_DEFENDER CLOSEST_DEFENDER_PLAYER_ID  \
43941         2       missed   Thompson, Klay                     202691   

       CLOSE_DEF_DIST  FGM  PTS     player_name player_id  
43941             6.6    0    0  nikola vucevic    202696  

[1 rows x 21 columns]

by looking at the data you would think that klay thompson is a good counter to nikola vucevic, unfortunetly nikola only took one shot against klay but he did miss

# James Harden defensive stats


In [ ]:
nv = missedshot[missedshot.CLOSEST_DEFENDER == 'Harden, James']

chart1 = alt.Chart(nv).mark_bar().encode(
    x=X('SHOT_DIST', bin=True),
    y='count()',
).properties(
    height=300,
    width=300
)

chart2 = alt.Chart(nv).mark_bar().encode(
    x=X('DRIBBLES', bin=True),
    y='count()'
).properties(
    height=300,
    width=300
)


chart3 = alt.Chart(nv).mark_bar().encode(
    x=X('CLOSE_DEF_DIST', bin=True),
    y='count()'
).properties(
    height=300,
    width=300
)
chart1 | chart2 | chart3

alt.HConcatChart(...)

In [ ]:
nv.mean(axis=0)

<ipython-input-20-9df0fb3fe1ce>:1: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  nv.mean(axis=0)


GAME_ID                       2.140044e+07
FINAL_MARGIN                 -2.644315e+00
SHOT_NUMBER                   6.688047e+00
PERIOD                        2.463557e+00
SHOT_CLOCK                    1.171793e+01
DRIBBLES                      1.632653e+00
TOUCH_TIME                    2.447813e+00
SHOT_DIST                     1.711691e+01
PTS_TYPE                      2.457726e+00
CLOSEST_DEFENDER_PLAYER_ID    2.019350e+05
CLOSE_DEF_DIST                4.538776e+00
FGM                           0.000000e+00
PTS                           0.000000e+00
player_id                     1.646015e+05
dtype: float64

dribbles avg: 1.6 <br>
shot distance: 17.1 <br>
avg distance to shooter: 4.5 <br>
as a defender James likes to keep shooters really far away from the hoop and close to him, he also likes to stop them from moving

In [ ]:
nvms = sl[sl.player_name == 'nikola vucevic']

In [ ]:
nvms = nvms[nvms.CLOSEST_DEFENDER == 'Harden, James']

In [ ]:
nvms

GAME_ID                     MATCHUP LOCATION  W  FINAL_MARGIN  \
43644  21400575  JAN 14, 2015 - ORL vs. HOU        H  W             7   
43645  21400575  JAN 14, 2015 - ORL vs. HOU        H  W             7   
43646  21400575  JAN 14, 2015 - ORL vs. HOU        H  W             7   
43647  21400575  JAN 14, 2015 - ORL vs. HOU        H  W             7   
43648  21400575  JAN 14, 2015 - ORL vs. HOU        H  W             7   
43652  21400575  JAN 14, 2015 - ORL vs. HOU        H  W             7   
43655  21400575  JAN 14, 2015 - ORL vs. HOU        H  W             7   

       SHOT_NUMBER  PERIOD GAME_CLOCK  SHOT_CLOCK  DRIBBLES  ...  SHOT_DIST  \
43644            6       2       6:00        20.1         0  ...        1.0   
43645            7       2       6:00        20.0         0  ...        1.0   
43646            8       2       5:23         8.2         0  ...       18.2   
43647            9       2       4:55        17.0         2  ...        9.1   
43648           10       2       3:07        24.0         0  ...        2.4   
43652           14       3       6:11        18.5         0  ...        4.5   
43655           17       4       0:48         3.8         0  ...        5.1   

       PTS_TYPE  SHOT_RESULT CLOSEST_DEFENDER CLOSEST_DEFENDER_PLAYER_ID  \
43644         2       missed    Harden, James                     201935   
43645         2       missed    Harden, James                     201935   
43646         2       missed    Harden, James                     201935   
43647         2         made    Harden, James                     201935   
43648         2         made    Harden, James                     201935   
43652         2         made    Harden, James                     201935   
43655         2         made    Harden, James                     201935   

       CLOSE_DEF_DIST  FGM  PTS     player_name player_id  
43644             0.8    0    0  nikola vucevic    202696  
43645             1.0    0    0  nikola vucevic    202696  
43646             4.8    0    0  nikola vucevic    202696  
43647             1.9    1    2  nikola vucevic    202696  
43648             1.6    1    2  nikola vucevic    202696  
43652             2.5    1    2  nikola vucevic    202696  
43655             2.3    1    2  nikola vucevic    202696  

[7 rows x 21 columns]

you can see in the above dataframe that when james harden does his job and keeps nikola far away, he misses his shot

# Jarrett Jack vs Avery Bradley

In [ ]:
nv = ms[ms.player_name == 'jarrett jack']

chart1 = alt.Chart(nv).mark_bar().encode(
    x=X('SHOT_DIST', bin=True),
    y='count()',
).properties(
    height=300,
    width=300
)

chart2 = alt.Chart(nv).mark_bar().encode(
    x=X('DRIBBLES', bin=True),
    y='count()'
).properties(
    height=300,
    width=300
)


chart3 = alt.Chart(nv).mark_bar().encode(
    x=X('CLOSE_DEF_DIST', bin=True),
    y='count()'
).properties(
    height=300,
    width=300
)
chart1 | chart2 | chart3

alt.HConcatChart(...)

Jarrett likes to takes shots from medium distance and likes to move a lot

In [ ]:
nv = missedshot[missedshot.CLOSEST_DEFENDER == 'Bradley, Avery']

chart1 = alt.Chart(nv).mark_bar().encode(
    x=X('SHOT_DIST', bin=True),
    y='count()',
).properties(
    height=300,
    width=300
)

chart2 = alt.Chart(nv).mark_bar().encode(
    x=X('DRIBBLES', bin=True),
    y='count()'
).properties(
    height=300,
    width=300
)


chart3 = alt.Chart(nv).mark_bar().encode(
    x=X('CLOSE_DEF_DIST', bin=True),
    y='count()'
).properties(
    height=300,
    width=300
)
chart1 | chart2 | chart3

alt.HConcatChart(...)

In [ ]:
nvms = sl[sl.player_name == 'jarrett jack']

In [ ]:
nvms = nvms[nvms.CLOSEST_DEFENDER == 'Bradley, Avery']

In [ ]:
nvms

GAME_ID                     MATCHUP LOCATION  W  FINAL_MARGIN  \
127747  21400527  JAN 07, 2015 - BKN vs. BOS        H  L            -8   
127748  21400527  JAN 07, 2015 - BKN vs. BOS        H  L            -8   
127753  21400527  JAN 07, 2015 - BKN vs. BOS        H  L            -8   
127833  21400432    DEC 26, 2014 - BKN @ BOS        A  W             2   
127837  21400432    DEC 26, 2014 - BKN @ BOS        A  W             2   
128061  21400006    OCT 29, 2014 - BKN @ BOS        A  L           -16   
128067  21400006    OCT 29, 2014 - BKN @ BOS        A  L           -16   
128068  21400006    OCT 29, 2014 - BKN @ BOS        A  L           -16   

        SHOT_NUMBER  PERIOD GAME_CLOCK  SHOT_CLOCK  DRIBBLES  ...  SHOT_DIST  \
127747            2       1       4:33        15.5         0  ...       24.0   
127748            3       1       2:33        15.4         0  ...       11.8   
127753            8       3       2:18         6.9         0  ...       24.0   
127833            6       3      11:01        14.0         1  ...       10.2   
127837           10       3       3:14        17.1         2  ...       14.9   
128061            2       2      10:10        19.0         0  ...        7.4   
128067            8       4       2:37         9.1         4  ...       18.3   
128068            9       4       0:12         NaN         5  ...        5.1   

        PTS_TYPE  SHOT_RESULT CLOSEST_DEFENDER CLOSEST_DEFENDER_PLAYER_ID  \
127747         3         made   Bradley, Avery                     202340   
127748         2       missed   Bradley, Avery                     202340   
127753         3       missed   Bradley, Avery                     202340   
127833         2       missed   Bradley, Avery                     202340   
127837         2         made   Bradley, Avery                     202340   
128061         2       missed   Bradley, Avery                     202340   
128067         2       missed   Bradley, Avery                     202340   
128068         2         made   Bradley, Avery                     202340   

        CLOSE_DEF_DIST  FGM  PTS   player_name player_id  
127747             6.8    1    3  jarrett jack    101127  
127748             3.9    0    0  jarrett jack    101127  
127753             4.6    0    0  jarrett jack    101127  
127833             2.1    0    0  jarrett jack    101127  
127837             3.9    1    2  jarrett jack    101127  
128061             2.7    0    0  jarrett jack    101127  
128067             3.0    0    0  jarrett jack    101127  
128068             2.3    1    2  jarrett jack    101127  

[8 rows x 21 columns]

as we can tell from the data jarrett missed a majority of shots in that 2 to 4 feet range that Avery is good at defending

# Conclusion
we can use data science to show the how the best shooters in the NBA can be countered by using other specific players, even if they are not from the same position <br>
matching a shooter to their perfect counter could yeild better defensive potential and I think this method of play is underutilized in basketball

In [ ]:
!jupyter nbconvert "Final.ipynb" --to pdf

[NbConvertApp] WARNING | pattern 'Final.ipynb' matched no files
This application is used to convert notebook files (*.ipynb)
        to various other formats.


Options
The options below are convenience aliases to configurable class-options,
as listed in the "Equivalent to" description-line of the aliases.
To see all configurable class-options for some <cmd>, use:
    <cmd> --help-all

--debug
    set log level to logging.DEBUG (maximize logging output)
    Equivalent to: [--Application.log_level=10]
--show-config
    Show the application's configuration (human-readable format)
    Equivalent to: [--Application.show_config=True]
--show-config-json
    Show the application's configuration (json format)
    Equivalent to: [--Application.show_config_json=True]
--generate-config
    generate default config file
    Equivalent to: [--JupyterApp.generate_config=True]
-y
    Answer yes to any questions instead of prompting.
    Equivalent to: [--JupyterApp.answer_yes=True]
--execute
    Execu